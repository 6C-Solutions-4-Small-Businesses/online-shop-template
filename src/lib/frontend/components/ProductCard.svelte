<script lang="ts">
    import {getAbbreviatedUnit} from '$lib/frontend/core/Helper'
    import {quintOut} from 'svelte/easing'
    import {scale} from 'svelte/transition'
    import ShoppingCartActions from '$lib/frontend/components/ShoppingCartActions.svelte'

    export let id: string
    export let name: string
    export let image: string
    export let regularPrice: number
    export let isSoldByQuantities: boolean
    export let unit: string | null
    export let salePrice: number | undefined | null = null
    export let percentage: number | null | undefined = null
    export let width: string
    export let height: string

    let expanded = false

    function getPriceFirstPart(price: number | undefined | null): string {
        const pricePart = price ? (price / 100).toString().split('.')[0] : null

        return pricePart ? pricePart : '00'
    }

    function getPriceSecondPart(price: number | undefined | null): string {
        const pricePart = price ? (price / 100).toFixed(2).toString().split('.')[1] : null

        return pricePart ? pricePart : '00'
    }

</script>

<div class={`product-box relative ${width} ${height} bg-white shadow-2xl`}
     on:mouseenter={()=> expanded = true}
     on:mouseleave={()=> expanded = false}
     role="none"
>
    <div
            class={`absolute flex w-12 h-12 top-0 text-white items-center ${percentage ? 'visible z-10' : 'invisible'}`}>
        <svg height="100" viewBox="0 0 490 490" width="100" x="0px" xml:space="preserve" y="0px">
      <polygon points="0,0 490,0 0,490"
               style={`fill: ${percentage && percentage <= 10 ? "#a67b5b" : percentage && percentage <= 50 ? "#fd7e14" : "#ff0000" };`}/>
    </svg>
        <div class="absolute -rotate-45 origin-center top-1 right-2 w-full text-[1rem] text-center font-semibold">
            -{percentage}%
        </div>
    </div>
    <img alt={name}
         class="product-image object-contain absolute h-full w-full top-0 p-4"
         src={image}>
    <div class="w-full absolute flex justify-end">
        <ShoppingCartActions
                image={image}
                name={name}
                productId={id}
                regularPrice={regularPrice}
                salePrice={salePrice}
        />
    </div>
    <div
            class={`absolute bottom-0 right-0 w-full flex justify-end`}>
        <div
                class={`filter backdrop-blur-sm bg-green-300/40 font-sans ${expanded ? 'w-full min-h-[4.2rem] justify-between' : 'w-28 h-[2.5rem] justify-center'} flex items-center text-lg`}
                role="region"
        >
            {#if expanded}
                <div class={`w-8/12 font-light text-xl px-2 text-wrap overflow-hidden text-overflow: ellipsis;`}
                     in:scale={{ duration: 500, opacity: 0.5, start: 0.5, easing: quintOut }}>
                    {name}
                </div>
            {/if}
            <div class={`${expanded ? 'w-4/12' : 'w-full'} flex flex-col`}>
                {#if (percentage && expanded)}
                    <div class="w-full flex-col items-center justify-center text-xl font-semibold">
                        ${getPriceFirstPart(salePrice)}
                        <sup>{getPriceSecondPart(salePrice)}</sup>
                        {#if (isSoldByQuantities)}
                            <span class="text-m block text-right pr-1"><sub>/ {getAbbreviatedUnit(unit)}</sub></span>
                        {/if}
                    </div>
                {/if}
                <div
                        class={`w-full items-center justify-center ${expanded ? "flex-col" : "flex"} ${expanded && percentage ? "line-through text-lg font-light" : "text-xl font-semibold"}`}>
                    <span class="block">${getPriceFirstPart(regularPrice)}<sup>{getPriceSecondPart(regularPrice)}</sup></span>
                    {#if (isSoldByQuantities)}
                        <span class="{expanded ? 'text-m block text-right pr-1' : 'text-sm'}"><sub>/ {getAbbreviatedUnit(unit)}</sub></span>
                    {/if}
                </div>
            </div>
        </div>
    </div>
</div>

<style lang="css">
    .product-image {
        aspect-ratio: 1;
    }

    .product-box:hover {
        box-shadow: 0 0 0 10px #16a34a inset;
    }
</style>
