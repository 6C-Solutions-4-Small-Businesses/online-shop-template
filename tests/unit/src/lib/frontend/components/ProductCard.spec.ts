import ProductCard from '$lib/frontend/components/ProductCard.svelte'
import {render} from '@testing-library/svelte'
import {describe, expect} from 'vitest'

describe('Product card', () => {
    it('should have match snapshot', () => {
        const {container} = render(ProductCard, {
            id: 'a-product-id',
            name: 'a-product-name',
            image: 'a-product-image',
            regularPrice: 250,
            salePrice: 150,
            isSoldByQuantities: false,
            unit: 'Centiliter',
        })

        expect(container).toMatchSnapshot()
    })
})

describe('Product card with promotion', () => {
    it('should have match snapshot', () => {
        const {container} = render(ProductCard, {
            id: 'a-product-id',
            name: 'a-product-name',
            image: 'a-product-image',
            regularPrice: 250,
            salePrice: 150,
            isSoldByQuantities: false,
            unit: 'Centiliter',
            percentage: 85
        })

        expect(container).toMatchSnapshot()
    })
})
