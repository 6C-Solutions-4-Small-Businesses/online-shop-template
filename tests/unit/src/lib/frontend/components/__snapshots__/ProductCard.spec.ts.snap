// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product card > should have match snapshot 1`] = `
<div>
  <div
    class="relative h-[326px] p-4 bg-white rounded-10 shadow flex-col justify-start items-start gap-[25px] inline-flex"
  >
     
    <div
      class="w-[238px] h-[200px] mt-10 self-center justify-between items-center flex"
    >
      <img
        alt="a-product-name"
        class="w-[238px] h-[200px] object-contain scale"
        src="a-product-image"
        width="160"
      />
    </div>
     
    <div
      class="w-full absolute flex justify-end top-0 right-0"
    >
      <div
        class="px-1 bg-white rounded-bl-[10px] rounded-tr-[10px] min-w-[40px] h-[50px] shadow justify-center items-center gap-2.5 inline-flex"
        role="region"
      >
         
        <button
          class="px-2 py-1 focus:outline-none"
        >
          <div
            class="h-6 w-6"
          >
            <svg
              fill="none"
              height="26"
              viewBox="0 0 18 26"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1 8.33334H2.77778L3.13333 10.1111M3.13333 10.1111H17L13.4444 17.2222H4.55556M3.13333 10.1111L4.55556 17.2222M4.55556 17.2222L2.51743 19.2604C1.95746 19.8204 2.35405 20.7778 3.14596 20.7778H13.4444M13.4444 20.7778C12.4626 20.7778 11.6667 21.5737 11.6667 22.5556C11.6667 23.5374 12.4626 24.3333 13.4444 24.3333C14.4263 24.3333 15.2222 23.5374 15.2222 22.5556C15.2222 21.5737 14.4263 20.7778 13.4444 20.7778ZM9.88889 1.66667V8.33334M6.33333 5.00001H13.4444M6.33333 22.5556C6.33333 23.5374 5.5374 24.3333 4.55556 24.3333C3.57372 24.3333 2.77778 23.5374 2.77778 22.5556C2.77778 21.5737 3.57372 20.7778 4.55556 20.7778C5.5374 20.7778 6.33333 21.5737 6.33333 22.5556Z"
                id="Add Button"
                stroke="#0A3B2F"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
     
    <div
      class="w-full flex-col justify-start items-start gap-2.5 flex"
    >
      <div
        class="w-full justify-between items-center flex"
      >
        <div
          class="text-accent pl-1.5 overflow-hidden max-h-[35px]"
        >
          a-product-name
        </div>
         
        <div
          class="text-primary text-15 flex ml-3"
        >
          CAD$
          2.50
           
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Product card with promotion > should have match snapshot 1`] = `
<div>
  <div
    class="relative h-[326px] p-4 bg-white rounded-10 shadow flex-col justify-start items-start gap-[25px] inline-flex"
  >
    <div
      class="absolute left-0 flex w-12 h-12 top-0 items-center justify-center rounded-tl-[10px] rounded-tr-none rounded-br-[20px] rounded-bl-none bg-promotion-high visible z-10"
    >
      <div
        class="text-white text-xs font-bold"
      >
        -
        85
        %
      </div>
    </div>
     
    <div
      class="w-[238px] h-[200px] mt-10 self-center justify-between items-center flex"
    >
      <img
        alt="a-product-name"
        class="w-[238px] h-[200px] object-contain scale"
        src="a-product-image"
        width="160"
      />
    </div>
     
    <div
      class="w-full absolute flex justify-end top-0 right-0"
    >
      <div
        class="px-1 bg-white rounded-bl-[10px] rounded-tr-[10px] min-w-[40px] h-[50px] shadow justify-center items-center gap-2.5 inline-flex"
        role="region"
      >
         
        <button
          class="px-2 py-1 focus:outline-none"
        >
          <div
            class="h-6 w-6"
          >
            <svg
              fill="none"
              height="26"
              viewBox="0 0 18 26"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1 8.33334H2.77778L3.13333 10.1111M3.13333 10.1111H17L13.4444 17.2222H4.55556M3.13333 10.1111L4.55556 17.2222M4.55556 17.2222L2.51743 19.2604C1.95746 19.8204 2.35405 20.7778 3.14596 20.7778H13.4444M13.4444 20.7778C12.4626 20.7778 11.6667 21.5737 11.6667 22.5556C11.6667 23.5374 12.4626 24.3333 13.4444 24.3333C14.4263 24.3333 15.2222 23.5374 15.2222 22.5556C15.2222 21.5737 14.4263 20.7778 13.4444 20.7778ZM9.88889 1.66667V8.33334M6.33333 5.00001H13.4444M6.33333 22.5556C6.33333 23.5374 5.5374 24.3333 4.55556 24.3333C3.57372 24.3333 2.77778 23.5374 2.77778 22.5556C2.77778 21.5737 3.57372 20.7778 4.55556 20.7778C5.5374 20.7778 6.33333 21.5737 6.33333 22.5556Z"
                id="Add Button"
                stroke="#0A3B2F"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
     
    <div
      class="w-full flex-col justify-start items-start gap-2.5 flex"
    >
      <div
        class="w-full justify-between items-center flex"
      >
        <div
          class="text-accent pl-1.5 overflow-hidden max-h-[35px]"
        >
          a-product-name
        </div>
         
        <div
          class="flex-col justify-start items-end gap-0.5 inline-flex"
        >
          <div
            class="flex items-end leading-normal flex-col"
          >
            <span
              class="text-primary text-[8px] line-through"
            >
              CAD$
              2.50
            </span>
             
            <div
              class="text-center text-error text-15 font-bold flex flex-row ml-3"
            >
              CAD$
              1.50
               
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
